rfcs/0137-nix-language-version.md

```diff
@@ -0,0 +1,678 @@
+  - Recursive (version bound):
+      - Issue warnings or errors on imported expressions written in prior versions (higher or equal than the version bound)
+  - Verbose mode. During evaluation:
+    - In non-verbose mode, issue a message once for each deprecated construct
+    - In verbose mode, issue a message for each occurrence
+
+   For example, this can be exposed as the following flags:
+
+   ```
+   --lang-no-warn=all
+   --lang-no-warn="url-literal let-body int-leading-zeroes non-attr-select"
+   --lang-error=all
+   --lang-error="url-literal let-body int-leading-zeroes non-attr-select"
+   --lang-warn-recursive=6
+   --lang-warn-verbose
+   ```
+
+   The naming may need some bikeshedding. For example, one could use the same syntax as with C-compilers (probably not though):
+
+   - `-Wall`
+   - `-Wno-`
+   - `-Werror=`
+
+   <details><summary>Arguments</summary>
+
+   - (+) Disallowing reuse avoids dealing with change of meaning across versions
+   - (+) Not considering experimental features simplifies their handling and is not required by them being exempt from compatibility guarantees.
+   - (-) Maintenance burden for everyone using these old constructs, or evaluating old revisions of Nixpkgs.
+   
+   </details>
+
+   <details><summary>Alternatives</summary>
+
+   - Opt-in warnings
+     - (-) Can't really make breaking changes as people won't be warned ahead of time
+   - No warnings, just errors
+     - (-) Doesn't offer a transition window to users
+     - (+) Easier to implement
+
+   </details>
+
+1. A the end of the evaluation, print statistics and explanations.
+   The specifics of displaying warnings and errors is up to implementation, but should include the symbolic name of the langauge construct in question.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+## Show the current Nix language version
+
+```console
+nix --language-version
+7
+```
+
+```console
+nix --supported-language-versions
+6
+7
+```
+
+## Version interoperability
+
+### Versioned built-ins can be passed across file boundaries
+
+```nix
+# a.nix
+version 6;
+builtins.langVersion
+```
+
+```nix
+# b.nix
+version 7;
+[ (import ./a.6.nix) builtins.langVersion ]
+```
+
+```console
+$ nix-instantiate --eval b.nix
+[ 6 7 ]
+```
+
+### Expressions are not forward compatible
+
+```nix
+# a.nix
+version 6;
+import ./b.nix
+```
+
+```nix
+# b.nix
+version 7;
+builtins.langVersion
+```
+
+```console
+$ nix-instantiate --eval a.nix
+error: unsupported Nix language version 7
+```
+
+
+### Best-effort interoperability
+
+```nix
+# a.nix
+version 6;
+{ increment }:
+increment 1
+```
+
+```nix
+# b.nix
+version 7;
+import ./a.nix { increment = x: x + 1 }
+```
+
+Since `increment` written in version 7 carries its own implementation with it, forcing it within an expression written in version 6 just works:
+
+```console
+$ nix-instantiate --eval b.nix
+2
+```
+
+### Pathological example
```

Comment:
Note that this means that only new value types can be added, but the representation can never be changed (unless the representation before and after are isomorphic). This means expressions can be versioned, but values never. The existence of thunks complicates that of course, since they are allowed versioned contents so to speak.

Normal values can never be changed with this proposal, e.g. we can't change strings from bytestrings to vectors of unicode code points or change the size of integer types.

Reply:
We *can* change representation values, but new language versions would have to provide the mappings, as you say. As you pointed out, due to this proposal aiming for very ambitious compatibility goals, it implies accordingly ambitious implementations.